File: packages/tui/test/support/owned-process.ts
import { mkdtemp, readFile, rm, stat, writeFile } from "node:fs/promises"
import { tmpdir } from "node:os"
import { join } from "node:path"
import { Worker } from "node:worker_threads"
import { registerTestProcess } from "./process-settlement"

interface ProcessResult { code: number; stdout: string; stderr: string }
interface ProcessOptions { cwd?: string; env?: Record<string, string>; timeoutMs: number; maxOutputBytes?: number }

/** A test directory remains owned until its Python supervisor proves group closure. */
export class TestProcessScope {
  readonly directory: string
  #pending: Promise<ProcessResult> | null = null
  #closing = false
  #unsettled = false

  private constructor(directory: string) { this.directory = directory }

  static async create(prefix: string): Promise<TestProcessScope> {
    return new TestProcessScope(await mkdtemp(join(tmpdir(), prefix)))
  }

  run(command: string[], options: ProcessOptions): Promise<ProcessResult> {
    if (this.#closing || this.#pending !== null || this.#unsettled) throw new Error("Test process scope is closed, busy, or UNSETTLED")
    const pending = this.#run(command, options)
    this.#pending = pending
    // Keep the actual physical promise reachable even if the test runner times out.
    void pending.finally(() => { if (this.#pending === pending) this.#pending = null }).catch(() => {})
    return pending
  }

  async #run(command: string[], options: ProcessOptions): Promise<ProcessResult> {
    const request = join(this.directory, "process.request.json")
    const result = join(this.directory, "process.result.json")
    const encoded = JSON.stringify({ command, cwd: options.cwd ?? process.cwd(), env: options.env ?? {},
      timeoutMs: options.timeoutMs, maxOutputBytes: options.maxOutputBytes ?? 1024 * 1024 })
    if (Buffer.byteLength(encoded) > 1024 * 1024) throw new Error("Oversized test process request")
    // One reusable result slot; no old acknowledgement may survive a new launch.
    await rm(result, { force: true })
    await writeFile(request, encoded)
    // This supervisor's own bounded deadline owns termination. Killing it could
    // abandon a native preload child; missing acknowledgement preserves evidence.
    this.#unsettled = true
    const registration = registerTestProcess()
    let protocolError: unknown = null
    // Bun's test VM automatically signals its subprocesses on test timeout,
    // including before Python installs its cooperative cancellation handler.
    // The owned worker VM keeps that startup handoff outside the test auto-killer.
    const code = await new Promise<number>((resolve) => {
      let bridgeCode = 125
      const worker = new Worker(new URL("./owned-process-worker.ts", import.meta.url), {
        workerData: { bridge: join(import.meta.dir, "owned-process.py"), request, result },
      })
      worker.on("message", (message: unknown) => {
        if (typeof message === "object" && message !== null && "started" in message &&
            typeof message.started === "number" && Number.isInteger(message.started) && message.started > 0) {
          try { registration.started(message.started) } catch (error) { protocolError = error }
        }
        if (typeof message === "number" && Number.isInteger(message)) bridgeCode = message
      })
      worker.on("error", () => { bridgeCode = 125 })
      worker.on("exit", (status) => resolve(status === 0 ? bridgeCode : 125))
    })
    if (protocolError !== null) throw new Error(`UNSETTLED test registration; retained ${this.directory}`, { cause: protocolError })
    let payload: unknown
    try {
      if ((await stat(result)).size > 6 * 1024 * 1024 + 8192) throw new Error("Missing bounded acknowledgement")
      payload = JSON.parse(await readFile(result, "utf8"))
      if (code !== 0 || typeof payload !== "object" || payload === null || !("settled" in payload) || payload.settled !== true) {
        const detail = typeof payload === "object" && payload !== null && "error" in payload && typeof payload.error === "string"
          ? payload.error.slice(0, 24 * 1024) : "Missing physical-settlement acknowledgement"
        throw new Error(detail)
      }
    } catch (error) {
      throw new Error(`UNSETTLED test process; retained ${this.directory}: ${String(error)}`, { cause: error })
    }
    registration.settled()
    this.#unsettled = false
    if ("error" in payload) throw new Error(String(payload.error))
    if (!("code" in payload) || !Number.isInteger(payload.code) || !("stdout" in payload) || typeof payload.stdout !== "string" ||
        !("stderr" in payload) || typeof payload.stderr !== "string") throw new Error("Invalid test process result")
    return { code: payload.code as number, stdout: payload.stdout, stderr: payload.stderr }
  }

  async close(): Promise<void> {
    this.#closing = true
    // Test timeout cleanup waits the same physical owner; no competing kill path.
    await this.#pending?.catch(() => {})
    if (this.#unsettled) throw new Error(`UNSETTLED test process; retained ${this.directory}`)
    await rm(this.directory, { recursive: true, force: true })
  }
}

File: packages/tui/test/support/owned-process.py
"""Run a test executable under the shared anchored process-group owner."""
from pathlib import Path
import json
import os
import sys

sys.path.insert(0, str(Path(__file__).resolve().parents[4] / "scripts"))
from perf_process import run_sample, require_sample_settlement
from owned_process_lifeline import ParentLifeline

MAX_REQUEST = 1024 * 1024
MAX_OUTPUT = 1024 * 1024


class CombinedOutput:
    def __init__(self, stream, limit):
        self.stream = stream
        self.limit = limit
        self.size = 0

    def write(self, chunk):
        self.stream.write(chunk[:max(0, self.limit - self.size)])
        self.stream.flush()
        self.size += len(chunk)
        if self.size > self.limit:
            raise ValueError("test child output exceeded its combined byte limit")

    def flush(self):
        self.stream.flush()


def run(request_path: Path, result_path: Path):
    if request_path.stat().st_size > MAX_REQUEST:
        raise ValueError("oversized test process request")
    request = json.loads(request_path.read_bytes())
    limit = request["maxOutputBytes"]
    timeout = request["timeoutMs"]
    if type(limit) is not int or not 0 < limit <= MAX_OUTPUT or type(timeout) is not int or not 0 < timeout <= 120_000:
        raise ValueError("invalid test process budgets")
    result = {"settled": True, "supervisor_pid": os.getpid()}
    log_path = request_path.with_suffix(".output.log")
    try:
        with log_path.open("wb") as log:
            completed = run_sample(
                request["command"], cwd=Path(request["cwd"]),
                env={**os.environ, **request["env"]}, timeout=timeout / 1000,
                output_limit=limit, log=CombinedOutput(log, limit),
            )
        result.update(code=completed.returncode, stdout=completed.stdout.decode("utf8", errors="replace"),
                      stderr=completed.stderr.decode("utf8", errors="replace"))
    except BaseException as error:
        with log_path.open("rb") as log:
            evidence = log.read(16 * 1024).decode("utf8", errors="replace")
        result["error"] = f"{type(error).__name__}: {error}"[:4096] + "\n" + evidence
    # Failed evidence is never a successful physical-closure acknowledgement.
    try:
        require_sample_settlement()
    except BaseException as error:
        result["settled"] = False
        result["error"] = result.get("error", "") + "\n" + str(error)[:4096]
    encoded = json.dumps(result, ensure_ascii=True).encode()
    if len(encoded) > MAX_OUTPUT * 6 + 8192:
        raise ValueError("oversized test process result")
    result_path.write_bytes(encoded)
    if not result["settled"]:
        raise SystemExit(125)


if __name__ == "__main__":
    try:
        with ParentLifeline(0):
            run(Path(sys.argv[1]), Path(sys.argv[2]))
    except BaseException:
        # Missing result is an explicit unproven lifetime; caller retains scratch.
        sys.exit(125)
